packages/react-native-gesture-handler/src/v3/hooks/events/useGestureStateChangeEvent.ts

```diff
@@ -0,0 +1,74 @@
+import { CALLBACK_TYPE } from '../../../handlers/gestures/gesture';
+import { compareTags, runWorklet } from '../utils';
+import { State } from '../../../State';
+import { Reanimated } from '../../../handlers/gestures/reanimatedWrapper';
+import { EventWithNativeEvent, StateChangeEvent } from '../../interfaces';
+import { GestureStateChangeEvent } from '../../../handlers/gestureHandlerCommon';
+
+export function useGestureStateChangeEvent(
+  handlerTag: number,
+  config: any,
+  shouldUseReanimated: boolean
+) {
+  const onGestureHandlerStateChange = (event: StateChangeEvent) => {
+    'worklet';
+
+    if (!compareTags(handlerTag, event)) {
+      return;
+    }
+
+    let oldState: State | undefined;
+    let state: State | undefined;
+
+    if (event.nativeEvent) {
+      oldState = (event as EventWithNativeEvent<GestureStateChangeEvent>)
+        .nativeEvent.oldState;
+      state = (event as EventWithNativeEvent<GestureStateChangeEvent>)
+        .nativeEvent.state;
+    } else {
+      oldState = (event as GestureStateChangeEvent).oldState;
+      state = (event as GestureStateChangeEvent).state;
+    }
+
+    if (oldState === State.UNDETERMINED && state === State.BEGAN) {
+      runWorklet(CALLBACK_TYPE.BEGAN, config, event);
+    } else if (
+      (oldState === State.BEGAN || oldState === State.UNDETERMINED) &&
+      state === State.ACTIVE
+    ) {
+      runWorklet(CALLBACK_TYPE.START, config, event);
+    } else if (oldState !== state && state === State.END) {
+      if (oldState === State.ACTIVE) {
+        runWorklet(CALLBACK_TYPE.END, config, event, true);
+      }
+      runWorklet(CALLBACK_TYPE.FINALIZE, config, event, true);
+    } else if (
+      (state === State.FAILED || state === State.CANCELLED) &&
+      state !== oldState
+    ) {
+      if (oldState === State.ACTIVE) {
+        runWorklet(CALLBACK_TYPE.END, config, event, false);
+      }
+      runWorklet(CALLBACK_TYPE.FINALIZE, config, event, false);
+    }
+  };
+
+  if (shouldUseReanimated) {
+    const handlers = {
+      onBegin: config.onBegin,
+      onStart: config.onStart,
+      onEnd: config.onEnd,
+      onFinalize: config.onFinalize,
+    };
```

Comment:
Can we use this instead of `config` in the event handler? Feels illegal to pass this to `useHandler` and then use a different object to call methods.

(the same in other files)

Reply:
Though I discussed it with Reanimated guys and they said that it should work, I changed it in [59e9c18](https://github.com/software-mansion/react-native-gesture-handler/pull/3617/commits/59e9c188aaa2d2bdcbeb50d05b9b86614ecd9ea9). Let me know what you think.